lib/AST/ASTPrinter.cpp

```diff
@@ -0,0 +1,659 @@
+      base_type = pty->getPointeeType();
+    else if (const auto bpy = base_type->getAs<clang::BlockPointerType>())
+      base_type = bpy->getPointeeType();
+    else if (const auto aty = clang::dyn_cast<clang::ArrayType>(base_type))
+      base_type = aty->getElementType();
+    else if (const auto fty = base_type->getAs<clang::FunctionType>())
+      base_type = fty->getReturnType();
+    else if (const auto vty = base_type->getAs<clang::VectorType>())
+      base_type = vty->getElementType();
+    else if (const auto pty = base_type->getAs<clang::ParenType>())
+      base_type = pty->desugar();
+    else
+      // This must be a syntax error.
+      break;
+  }
+  return base_type;
+}
+
+static clang::QualType GetDeclType(clang::Decl *decl) {
+  if (auto tdd = clang::dyn_cast<clang::TypedefNameDecl>(decl)) {
+    return tdd->getUnderlyingType();
+  }
+
+  if (auto vd = clang::dyn_cast<clang::ValueDecl>(decl)) {
+    return vd->getType();
+  }
+
+  return clang::QualType();
+}
+
+}  // namespace
+
+void DeclTokenizer::Indent() {
+  for (auto i{0U}; i < indent; ++i) {
+    out.push_back(Token::CreateIndent());
+  }
+}
+
+void DeclTokenizer::PrintPragmas(clang::Decl *decl) {}
+
+void DeclTokenizer::PrintAttributes(clang::Decl *decl) {}
+
+void DeclTokenizer::VisitVarDecl(clang::VarDecl *vdecl) {
+  PrintPragmas(vdecl);
+
+  clang::QualType vtype;
+  if (vdecl->getTypeSourceInfo()) {
+    vtype = vdecl->getTypeSourceInfo()->getType();
+  }
+
+  clang::StorageClass sc{vdecl->getStorageClass()};
+  if (sc != clang::SC_None) {
+    std::string scs_str{clang::VarDecl::getStorageClassSpecifierString(sc)};
+    out.push_back(Token::CreateMisc(scs_str));
+    out.push_back(Token::CreateSpace());
+  }
+
+  switch (vdecl->getTSCSpec()) {
+    case clang::TSCS_unspecified:
+      break;
+
+    case clang::TSCS___thread:
+      out.push_back(Token::CreateMisc("__thread"));
+      out.push_back(Token::CreateSpace());
+      break;
+
+    case clang::TSCS__Thread_local:
+      out.push_back(Token::CreateMisc("_Thread_local"));
+      out.push_back(Token::CreateSpace());
+      break;
+
+    case clang::TSCS_thread_local:
+      out.push_back(Token::CreateMisc("thread_local"));
+      out.push_back(Token::CreateSpace());
+      break;
+  }
+
+  if (vdecl->isModulePrivate()) {
+    out.push_back(Token::CreateMisc("__module_private__"));
+    out.push_back(Token::CreateSpace());
+  }
+
+  std::string buf{""};
+  llvm::raw_string_ostream ss(buf);
+  vtype.print(ss, unit.getASTContext().getPrintingPolicy(), vdecl->getName(),
+              /*Indentation=*/0U);
+
+  out.push_back(Token::CreateDecl(vdecl, ss.str()));
+
+  if (auto init = vdecl->getInit()) {
+    if (vdecl->getInitStyle() == clang::VarDecl::CInit) {
+      out.push_back(Token::CreateSpace());
+      out.push_back(Token::CreateMisc("="));
+      out.push_back(Token::CreateSpace());
+    } else {
+      LOG(FATAL) << "Invalid initialization style.";
+    }
+
+    StmtTokenizer(out, unit).Visit(init);
```

Comment:
Is it possible to associate the `vdecl` with all tokens introduced by this `stmt` (assuming those tokens aren't already tied to another decl)?  In that case, the nodes are associated with a decl and a stmt.

Reply:
> No, this is not directly possible. 

Can you store the current index into the token output list, then iterate over the tokens added into the list by the call to `Visit`, then assign `vdecl`? This type of thing could probably wrapped in some kind of context manager. If you did the 1:N mapping, then you could have the associated modelled as a vector (a stack really) and then that would naturally coincide with the recursive calls to `.Visit(...)`.

Perhaps even doing everything via context management may prove the easiest solution, because you wouldn't need to manually assign decls/stmts/types to tokens!!